In [1]:
import pandas

# Introductie

In dit notebook wordt data uit twee csv files samengevoegd en dit resulteert in de file: gemeentedata_merged.csv<br>
Het gaat om:
- gegevens_gemeenten_2021.csv<br>
&nbsp;&nbsp;&nbsp;&nbsp; Bevat gegevens m.b.t. o.a. woningprijs, bevolkingssamenstelling, inkomen, werkloosheid, criminaliteit per gemeente in Nederland
- Uitslag_alle_gemeenten_TK20210317.csv<br>
&nbsp;&nbsp;&nbsp;&nbsp; Bevat gegevens m.b.t. de verkiezingsuitslag van de Tweede Kamer verkiezingen van 2021 per gemeente in Nederland


# Importeren data

In [2]:
data_diverse = pandas.read_csv('./data/gegevens_gemeenten_2021.csv', sep=';')
election_data = pandas.read_csv('./data/Uitslag_alle_gemeenten_TK20210317.csv', sep=';')

In [3]:
data_diverse = data_diverse.sort_values('Gemeenten')
election_data = election_data.sort_values('RegioNaam')

In [4]:
len(data_diverse)

344

In [5]:
len(election_data)

355

In [6]:
data_diverse.head(5)

Gemeenten  Vraagprijs aangeboden woningen|2021  \
252  's-Hertogenbosch                               488020   
0         Aa en Hunze                               469107   
1            Aalsmeer                               658553   
2              Aalten                               347721   
3       Achtkarspelen                               357519   

     Bevolking totaal|2021 Leeftijd bevolking|2021  Bevolkingsdichtheid|2021  \
252                 155490                      42                      1414   
0                    25399                    47,3                        92   
1                    31991                    42,2                      1590   
2                    27120                    44,5                       281   
3                    27900                    42,3                       273   

     Migratieachtergrond - Nederlandse achtergrond|2021  \
252                                             122514    
0                                                23719    
1                                                25117    
2                                                23997    
3                                                26631    

     Migratieachtergrond - Westers|2021  \
252                               14291   
0                                  1062   
1                                  3708   
2                                  2159   
3                                   666   

     Migratieachtergrond - Niet-westers|2021  \
252                                    18685   
0                                        618   
1                                       3166   
2                                        964   
3                                        603   

     Migratieachtergrond - Totaal|2021  Totaal misdrijven|2021  \
252                              32976                    8398   
0                                 1680                     605   
1                                 6874                    1008   
2                                 3123                     496   
3                                 1269                     631   

     Misdrijven - Diefstal/inbraak woning|2021 Werkloosheidspercentage|2021  \
252                                      251.0                          3,2   
0                                         23.0                          3,4   
1                                         34.0                          3,8   
2                                         15.0                          2,9   
3                                          8.0                          3,9   

    Besteedbaar inkomen per huishouden|2019  
252                                    45,1  
0                                      47,1  
1                                        55  
2                                      44,1  
3                                      40,4

In [7]:
election_data.head(5)

RegioNaam RegioCode  AmsterdamseCode    OuderRegioNaam  \
258     's-Gravenhage     G0518            11434     's-Gravenhage   
259  's-Hertogenbosch     G0796            10054  's-Hertogenbosch   
0         Aa en Hunze     G1680            10787             Assen   
1            Aalsmeer     G0358            11264           Haarlem   
2              Aalten     G0197            11046            Arnhem   

    OuderRegioCode  Kiesgerechtigden  Opkomst  OngeldigeStemmen  \
258            K12            449253   319724              1423   
259            K18            121353    95871               186   
0               K3             20921    17444                32   
1              K10             23249    19563                23   
2               K7             21386    17215                18   

     BlancoStemmen  GeldigeStemmen  ...  OPRECHT  JEZUS LEEFT  \
258            530          317771  ...    175.0          NaN   
259            153           95532  ...     32.0         18.0   
0               25           17387  ...      9.0          3.0   
1               35           19505  ...     11.0          NaN   
2               16           17181  ...      4.0         12.0   

     DE FEESTPARTIJ (DFP)  U-Buntu Connected Front  Vrij en Sociaal Nederland  \
258                   NaN                     87.0                      139.0   
259                  73.0                      NaN                        NaN   
0                     NaN                      NaN                        NaN   
1                    12.0                      2.0                        5.0   
2                     NaN                      1.0                        NaN   

     Partij van de Eenheid  Wij zijn Nederland  Partij voor de Republiek  \
258                   96.0                 NaN                       NaN   
259                   20.0                 NaN                       NaN   
0                      NaN                 NaN                       NaN   
1                      NaN                 2.0                       NaN   
2                      NaN                 NaN                       NaN   

     Modern Nederland  De Groenen  
258               NaN         NaN  
259               NaN         NaN  
0                 2.0         NaN  
1                 NaN         NaN  
2                 NaN         NaN  

[5 rows x 47 columns]

# Vergelijking gemeentenamen tussen de twee data sheets

In [8]:
result_inner_join = pandas.merge(data_diverse.iloc[:, [0]], election_data.iloc[:, [0]],
                                 how='inner',
                                 left_on='Gemeenten', right_on='RegioNaam')

In [9]:
len(result_inner_join)

338

We zien dat 338 namen hetzelfde zijn.

In [10]:
result_left_join = pandas.merge(data_diverse.iloc[:, [0]], election_data.iloc[:, [0]],
                                how='left',
                                left_on='Gemeenten', right_on='RegioNaam')

In [11]:
result_left_join[(result_left_join.RegioNaam != result_left_join.Gemeenten)]

Gemeenten RegioNaam
30      Bergen (L.)       NaN
31     Bergen (NH.)       NaN
67         Den Haag       NaN
72    Dijk en Waard       NaN
157  Land van Cuijk       NaN
177       Maashorst       NaN

- We zien dat 6 rijen in data_diverse een gemeentenaam hebben die niet gematched kan worden op een gemeentenaam in election_data

In [12]:
result_right_join = pandas.merge(data_diverse.iloc[:, [0]], election_data.iloc[:, [0]],
                    how='right',
                    left_on='Gemeenten', right_on='RegioNaam')

In [13]:
result_right_join[(result_right_join.Gemeenten != result_right_join.RegioNaam)]

Gemeenten            RegioNaam
0         NaN        's-Gravenhage
28        NaN             Beemster
32        NaN               Bergen
33        NaN               Bergen
45        NaN              Bonaire
49        NaN              Boxmeer
63        NaN                Cuijk
112       NaN                Grave
128       NaN        Heerhugowaard
162       NaN              Landerd
165       NaN            Langedijk
193       NaN  Mill en Sint Hubert
254       NaN                 Saba
261       NaN        Sint Anthonis
262       NaN       Sint Eustatius
290       NaN                 Uden
322       NaN                Weesp

- We zien dat 17 rijen in election_data een gemeentenaam hebben die niet gematched kan worden op een gemeentenaam in data_diverse

## Verklaring van de waargenomen verschillen

- Sommige namen konden (net) niet gematched worden, wegens een (klein) verschil in naamgeving

Bergen (L.)<br>
Bergen

Bergen (NH.)<br>
Bergen

Den Haag<br>
's-Gravenhage

- Bijzondere gemeenten

Bonaire<br>
Saba<br>
Sint Eustatius


- Opgeheven gemeenten die werden ondergebracht bij een andere gemeente

Beemster (gefuseerd met Purmerend)<br>
Weesp (gefuseerd met Amsterdam)<br>

- Samengevoegde gemeenten

**Dijk en Waard**<br>
Heerhugowaard<br>
Langedijk<br>

**Land van Cuijk**<br>
Boxmeer<br>
Cuijk<br>
Grave<br>
Mill en Sint Hubert<br>
Sint Anthonis<br>

**Maashorst**<br>
Landerd<br>
Uden<br>

- Wat betreft de gemeentelijke herindeling:<br><br>De gegevens m.b.t. 2021 werden gedownload in 2022, en voor de gegevens
die gedownload werden van de website www.waarstaatjegemeente.nl/jive geldt dat standaard uitgegaan wordt van
de actuele gemeentelijke indeling, ook voor gegevens uit voorgaande jaren.<br>
De gemeentelijke herindeling na de verkiezingen van 2021 verklaart de bovengenoemde verschillen.<br><br>
Op: https://www.cbs.nl/nl-nl/onze-diensten/methoden/classificaties/overig/gemeentelijke-indelingen-per-jaar lezen we: 'Per 1 januari 2022 is het aantal gemeenten door gemeentelijke herindelingen afgenomen tot 345. En per 24 maart 2022 tot 344 gemeenten.'<br><br>
Om e.e.a. te corrigeren dienen we de verkiezingsdata te converteren, zodat ook hier wordt uitgegaan van de gemeentelijke indeling, zoals die gold begin oktober 2022.

# Opschonen van de data

### Deel I

Eerst schonen we de kolomnamen op:

- In beide sheets de kolom met gemeentenamen de kolomnaam 'Gemeente' geven

In [14]:
data_diverse.rename(columns={'Gemeenten': 'Gemeente'}, inplace=True)
election_data.rename(columns={'RegioNaam': 'Gemeente'}, inplace=True)

- in de sheet data_diverse diverse kolomnamen aanpassen

In [15]:
data_diverse.rename(columns={'Vraagprijs aangeboden woningen|2021': 'Gemiddelde vraagprijs aangeboden woningen',
                             'Bevolking totaal|2021': 'Bevolking totaal',
                             'Leeftijd bevolking|2021': 'Gemiddelde leeftijd bevolking',
                             'Bevolkingsdichtheid|2021': 'Bevolkingsdichtheid',
                             'Totaal misdrijven|2021': 'Misdrijven totaal',
                             'Misdrijven - Diefstal/inbraak woning|2021': 'Diefstal/inbraak',
                             'Werkloosheidspercentage|2021': 'Werkloosheidspercentage',
                             'Migratieachtergrond - Nederlandse achtergrond|2021': 'Autochtonen totaal',
                             'Migratieachtergrond - Westers|2021': 'Allochtonen Westers',
                             'Migratieachtergrond - Niet-westers|2021': 'Allochtonen Niet Westers',
                             'Migratieachtergrond - Totaal|2021': 'Allochtonen totaal',
                             }, inplace=True)

- In de sheet election_data elke partijnaam vervangen door een bondigere variant

In [16]:
election_data.rename(columns={'PVV (Partij voor de Vrijheid)': 'PVV',
                              'SP (Socialistische Partij)': 'SP',
                              'Partij van de Arbeid (P.v.d.A.)': 'PvdA',
                              'Forum voor Democratie': 'FvD',
                              'Partij voor de Dieren': 'PvdD',
                              'Staatkundig Gereformeerde Partij (SGP)': 'SGP',
                              'Trots op Nederland (TROTS)': 'TROTS',
                              'Blanco (Zeven, A.J.L.B.)': 'Blanco',
                              'LP (Libertaire Partij)': 'LP',
                              'DE FEESTPARTIJ (DFP)': 'DFP',
                              'Partij van de Eenheid': 'PvdE',
                              'Partij voor de Republiek': 'PvdR',
                              }, inplace=True)

- Enkele correcties m.b.t. het format van waarden: een komma vervangen door een punt

In [17]:
for colname in ['Gemiddelde leeftijd bevolking',
                'Werkloosheidspercentage',
                ]:

    data_diverse[colname] = data_diverse[colname].apply(lambda value: str(value).replace(',', '.')).apply(float)

- In geval van 'Besteedbaar inkomen per huishouden|2019' ligt het iets complexer, omdat we hier ook met missing data
te maken hebben.<br>
    * Sommige cellen hebben een waarde '?'
    * We converteren deze eerst naar '0', daarna converteren we alle waarden in de gehele kolom naar float
    * Vervolgens berekenen we het gemiddelde en zetten we alle cellen, met waarde 0 op dit gemiddelde.

In [18]:
filter_qm = data_diverse['Besteedbaar inkomen per huishouden|2019'].isin(['?'])

In [19]:
data_diverse[filter_qm]

Gemeente  Gemiddelde vraagprijs aangeboden woningen  \
12           Ameland                                     554807   
234        Renswoude                                     518110   
245        Rozendaal                                     925931   
250  Schiermonnikoog                                     520548   
271     Terschelling                                     493591   
298         Vlieland                                     385608   

     Bevolking totaal  Gemiddelde leeftijd bevolking  Bevolkingsdichtheid  \
12               3746                           44.2                   63   
234              5556                           37.8                  302   
245              1726                           45.1                   62   
250               931                           49.9                   23   
271              4870                           45.1                   57   
298              1194                           43.8                   30   

     Autochtonen totaal  Allochtonen Westers  Allochtonen Niet Westers  \
12                 3439                  239                        68   
234                5092                  275                       189   
245                1458                  168                       100   
250                 830                   77                        24   
271                4500                  273                        97   
298                 997                  171                        26   

     Allochtonen totaal  Misdrijven totaal  Diefstal/inbraak  \
12                  307                 95               2.0   
234                 464                 96               6.0   
245                 268                 46               1.0   
250                 101                 33               NaN   
271                 370                170               3.0   
298                 197                 80               1.0   

     Werkloosheidspercentage Besteedbaar inkomen per huishouden|2019  
12                       3.4                                       ?  
234                      2.7                                       ?  
245                      2.8                                       ?  
250                      3.2                                       ?  
271                      3.1                                       ?  
298                      3.2                                       ?

In [20]:
data_diverse['Besteedbaar inkomen per huishouden|2019'] = data_diverse['Besteedbaar inkomen per huishouden|2019'].apply(lambda value: value.replace('?', '0'))
data_diverse['Besteedbaar inkomen per huishouden|2019'] = data_diverse['Besteedbaar inkomen per huishouden|2019'].apply(lambda value: value.replace(',', '.'))
data_diverse['Besteedbaar inkomen per huishouden|2019'] = data_diverse['Besteedbaar inkomen per huishouden|2019'].apply(lambda value: float(value))

In [21]:
mean = round(data_diverse['Besteedbaar inkomen per huishouden|2019'].mean(), 1)

In [22]:
data_diverse['Besteedbaar inkomen per huishouden|2019'] = data_diverse['Besteedbaar inkomen per huishouden|2019'].apply(lambda value: mean if value == 0 else value)

- Nacontrole

In [23]:
filter_mean = data_diverse['Besteedbaar inkomen per huishouden|2019'].isin([mean])

In [24]:
data_diverse[filter_mean]

Gemeente  Gemiddelde vraagprijs aangeboden woningen  \
12           Ameland                                     554807   
45            Boxtel                                     515983   
126      Hellendoorn                                     354764   
231           Raalte                                     351904   
234        Renswoude                                     518110   
245        Rozendaal                                     925931   
250  Schiermonnikoog                                     520548   
271     Terschelling                                     493591   
298         Vlieland                                     385608   

     Bevolking totaal  Gemiddelde leeftijd bevolking  Bevolkingsdichtheid  \
12               3746                           44.2                   63   
45              32973                           44.0                  477   
126             35932                           43.7                  261   
231             37911                           44.2                  222   
234              5556                           37.8                  302   
245              1726                           45.1                   62   
250               931                           49.9                   23   
271              4870                           45.1                   57   
298              1194                           43.8                   30   

     Autochtonen totaal  Allochtonen Westers  Allochtonen Niet Westers  \
12                 3439                  239                        68   
45                27444                 2760                      2769   
126               33366                 1454                      1112   
231               35287                 1324                      1300   
234                5092                  275                       189   
245                1458                  168                       100   
250                 830                   77                        24   
271                4500                  273                        97   
298                 997                  171                        26   

     Allochtonen totaal  Misdrijven totaal  Diefstal/inbraak  \
12                  307                 95               2.0   
45                 5529               1283              43.0   
126                2566                732              31.0   
231                2624                758              20.0   
234                 464                 96               6.0   
245                 268                 46               1.0   
250                 101                 33               NaN   
271                 370                170               3.0   
298                 197                 80               1.0   

     Werkloosheidspercentage  Besteedbaar inkomen per huishouden|2019  
12                       3.4                                     46.5  
45                       3.5                                     46.5  
126                      2.9                                     46.5  
231                      2.8                                     46.5  
234                      2.7                                     46.5  
245                      2.8                                     46.5  
250                      3.2                                     46.5  
271                      3.1                                     46.5  
298                      3.2                                     46.5

- Alle overige ontbrekende waarden geven we een waarde 0

In [25]:
election_data = election_data.fillna(0)
data_diverse = data_diverse.fillna(0)

- Het betreft de inbraak gegevens van Schiermonnikoog<br>
- Verder alle cellen in de verkiezingsdata, waar verzuimd werd een 0 in te voeren (0 stemmen)

### Deel II

- Een aantal gemeenten moeten qua naam gelijk getrokken worden

In [26]:
for name, regiocode, new_name in [
                ['\'s-Gravenhage', 'G0518', 'Den Haag'],
                ['Bergen', 'G0373', 'Bergen (NH.)'],
                ['Bergen', 'G0893', 'Bergen (L.)'],
                      ]:

    filter1 = election_data['Gemeente'].isin([name])
    filter2 = election_data['RegioCode'].isin([regiocode])
    index = election_data[filter1][filter2].index
    election_data.loc[index, 'Gemeente'] = new_name

<ipython-input-26-5a47c70f2024>:9: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  index = election_data[filter1][filter2].index


- De bijzondere gemeenten komen te vervallen in election_data

In [27]:
election_data = election_data.drop(election_data[(election_data.Gemeente.isin(['Bonaire', 'Saba', 'Sint Eustatius']))].index)

- Nu het probleem oplossen met de samengevoegde gemeenten:<br>
In election_data dienen<br>
    * de nieuwe gemeentenamen toegevoegd te worden<br>
    * de uitgefaseerde gemeentenamen verwijderd te worden<br>
    * de aantallen van de uitgefaseerde gemeenten toegevoegd te worden aan de gemeente waarin ze werden opgenomen

In [28]:
conversion_table = {
    'Dijk en Waard': ['Heerhugowaard', 'Langedijk'],
    'Land van Cuijk': ['Boxmeer', 'Cuijk', 'Grave', 'Mill en Sint Hubert', 'Sint Anthonis'],
    'Maashorst': ['Landerd', 'Uden'],
    'Purmerend': ['Purmerend', 'Beemster'],
    'Amsterdam': ['Amsterdam', 'Weesp'],
}

In [29]:
def find_new_municipalities(conversion_table):
    return [municipality for municipality in conversion_table if municipality not in conversion_table[municipality]]


def find_updated_municipalities(conversion_table):
    return [municipality for municipality in conversion_table if municipality in conversion_table[municipality]]


def find_phased_out_municipalities(conversion_table, updated_municipalities):

    phased_out_municipalities = []

    for municipality_ in conversion_table:
        for municipality in conversion_table[municipality_]:
            if municipality not in updated_municipalities:
                phased_out_municipalities.append(municipality)

    return phased_out_municipalities


def add_new_municipalities(new_municipalities, conversion_table, election_data):
    
    for new_municipality in new_municipalities:
        phased_out_municipalities = conversion_table[new_municipality]
    
        municipality_to_add = election_data[(election_data.Gemeente.isin(phased_out_municipalities))].sum()
        municipality_to_add.Gemeente = new_municipality
        election_data = election_data.append(municipality_to_add, ignore_index=True)

        indices_rows2drop = election_data[(election_data.Gemeente.isin(phased_out_municipalities))].index
        election_data = election_data.drop(indices_rows2drop)

    return election_data


def update_municipalities(updated_municipalities, conversion_table, election_data):
    
    for updated_municipality in updated_municipalities:
        
        municipalities2merge = conversion_table[updated_municipality]

        municipality_to_add = election_data[(election_data.Gemeente.isin(municipalities2merge))].sum()
        election_data = election_data.drop(election_data[(election_data.Gemeente.isin(municipalities2merge))].index)
        municipality_to_add.Gemeente = updated_municipality
        election_data = election_data.append(municipality_to_add, ignore_index=True)
    
    return election_data


new_municipalities = find_new_municipalities(conversion_table)
updated_municipalities = find_updated_municipalities(conversion_table)
phased_out_municipalities = find_phased_out_municipalities(conversion_table, updated_municipalities)
    
election_data = add_new_municipalities(new_municipalities, conversion_table, election_data)
election_data = update_municipalities(updated_municipalities, conversion_table, election_data)

# Nacontrole

In [30]:
len(data_diverse)

344

In [31]:
len(election_data)

344

In [32]:
result_inner_join = pandas.merge(data_diverse.iloc[:, [0]], election_data.iloc[:, [0]],
                                 how='inner',
                                 left_on='Gemeente', right_on='Gemeente')

In [33]:
len(result_inner_join)

344

# Samenvoegen data

In [34]:
gemeentedata = pandas.merge(data_diverse, election_data, on='Gemeente')

In [35]:
gemeentedata.head(4)

Gemeente  Gemiddelde vraagprijs aangeboden woningen  \
0  's-Hertogenbosch                                     488020   
1       Aa en Hunze                                     469107   
2          Aalsmeer                                     658553   
3            Aalten                                     347721   

   Bevolking totaal  Gemiddelde leeftijd bevolking  Bevolkingsdichtheid  \
0            155490                           42.0                 1414   
1             25399                           47.3                   92   
2             31991                           42.2                 1590   
3             27120                           44.5                  281   

   Autochtonen totaal  Allochtonen Westers  Allochtonen Niet Westers  \
0              122514                14291                     18685   
1               23719                 1062                       618   
2               25117                 3708                      3166   
3               23997                 2159                       964   

   Allochtonen totaal  Misdrijven totaal  ...  OPRECHT  JEZUS LEEFT   DFP  \
0               32976               8398  ...     32.0         18.0  73.0   
1                1680                605  ...      9.0          3.0   0.0   
2                6874               1008  ...     11.0          0.0  12.0   
3                3123                496  ...      4.0         12.0   0.0   

  U-Buntu Connected Front  Vrij en Sociaal Nederland  PvdE Wij zijn Nederland  \
0                     0.0                        0.0  20.0                0.0   
1                     0.0                        0.0   0.0                0.0   
2                     2.0                        5.0   0.0                2.0   
3                     1.0                        0.0   0.0                0.0   

   PvdR  Modern Nederland  De Groenen  
0   0.0               0.0         0.0  
1   0.0               2.0         0.0  
2   0.0               0.0         0.0  
3   0.0               0.0         0.0  

[4 rows x 59 columns]

In [36]:
gemeentedata.columns

Index(['Gemeente', 'Gemiddelde vraagprijs aangeboden woningen',
       'Bevolking totaal', 'Gemiddelde leeftijd bevolking',
       'Bevolkingsdichtheid', 'Autochtonen totaal', 'Allochtonen Westers',
       'Allochtonen Niet Westers', 'Allochtonen totaal', 'Misdrijven totaal',
       'Diefstal/inbraak', 'Werkloosheidspercentage',
       'Besteedbaar inkomen per huishouden|2019', 'RegioCode',
       'AmsterdamseCode', 'OuderRegioNaam', 'OuderRegioCode',
       'Kiesgerechtigden', 'Opkomst', 'OngeldigeStemmen', 'BlancoStemmen',
       'GeldigeStemmen', 'VVD', 'D66', 'PVV', 'CDA', 'SP', 'PvdA',
       'GROENLINKS', 'FvD', 'PvdD', 'ChristenUnie', 'Volt', 'JA21', 'SGP',
       'DENK', '50PLUS', 'BBB', 'BIJ1', 'CODE ORANJE', 'NIDA', 'Splinter',
       'Piratenpartij', 'JONG', 'TROTS', 'Lijst Henk Krol', 'NLBeter',
       'Blanco', 'LP', 'OPRECHT', 'JEZUS LEEFT', 'DFP',
       'U-Buntu Connected Front', 'Vrij en Sociaal Nederland', 'PvdE',
       'Wij zijn Nederland', 'PvdR', 'Modern Nederland

# Export to csv

In [37]:
gemeentedata.to_csv('./data/gemeentedata_merged.csv', index=False, sep=';')